vm/src/builtins/str.rs

```diff
@@ -928,9 +928,13 @@ impl PyStr {
     }
 
     #[pymethod]
-    fn join(&self, iterable: ArgIterable<PyStrRef>, vm: &VirtualMachine) -> PyResult<String> {
-        let iter = iterable.iter(vm)?;
-        self.as_str().py_join(iter)
+    fn join(&self, iterable: ArgIterable<PyStrRef>, vm: &VirtualMachine) -> PyResult<PyStrRef> {
+        let mut iter = iterable.iter(vm)?;
+
+        match iter.size_hint().0 {
```

Comment:
`size_hint` is not a length itself. I think retrieving all items to a `Vec` and checking the size is the correct way.